### PR TITLE
User interface layer

### PR DESCRIPTION
**Pull request by Axinalis**

### CODE REVIEW

File: JpaLayer/src/main/java/com/axinalis/config/KafkaConsumersConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+    @Primary
+    @Bean
+    public Map<String, Object> consumerConfiguration(){
+        Map<String, Object> configuration = new HashMap<>();
+
+        configuration.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, KAFKA_BROKER);
+        configuration.put(ConsumerConfig.GROUP_ID_CONFIG, GROUP_ID);
+        configuration.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
+        configuration.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
```
Comment (nolik) [inline JpaLayer/src/main/java/com/axinalis/config/KafkaConsumersConfiguration.java]: advice to use JSON serialization for value.

File: UserInterfaceLayer/src/main/java/com/axinalis/config/KafkaConsumersConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import org.springframework.kafka.core.DefaultKafkaConsumerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+
+@Configuration
+public class KafkaConsumersConfiguration {
+
+    @Value("${KAFKA_BROKER}")
+    private String KAFKA_BROKER;
```
Comment (nolik) [inline UserInterfaceLayer/src/main/java/com/axinalis/config/KafkaConsumersConfiguration.java]: please follow naming conventions (i.e. as far it's not constants).

File: UserInterfaceLayer/src/main/java/com/axinalis/config/KafkaConsumersConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+    @Primary
+    @Bean
+    public Map<String, Object> consumerConfiguration(){
+        Map<String, Object> configuration = new HashMap<>();
+
+        configuration.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, KAFKA_BROKER);
+        configuration.put(ConsumerConfig.GROUP_ID_CONFIG, GROUP_ID);
+        configuration.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
+        configuration.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
```
Comment (nolik) [inline UserInterfaceLayer/src/main/java/com/axinalis/config/KafkaConsumersConfiguration.java]: same here. you may specify JSON deserializaer.

File: UserInterfaceLayer/src/main/java/com/axinalis/consumer/TransactionsConsumer.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    private final List<Transaction> list;
+
+    public TransactionsConsumer(@Autowired ObjectMapper mapper) {
+        this.mapper = mapper;
+        list = new LinkedList<>();
+    }
+
+    @KafkaListener(topics = "${TRANSACTIONS_TOPIC}", groupId = "${GROUP_ID}")
+    public void listen(String message) throws JsonProcessingException {
+        synchronized (list){
```
Comment (nolik) [inline UserInterfaceLayer/src/main/java/com/axinalis/consumer/TransactionsConsumer.java]: worth just log.

File: JpaLayer/pom.xml [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>com.axinalis</groupId>
+  <artifactId>JpaLayer</artifactId>
+  <version>1.0</version>
+  <name>JpaLayer</name>
+  <url>http://maven.apache.org</url>
+
+  <properties>
+    <maven.compiler.source>11</maven.compiler.source>
```
Comment (nolik) [inline JpaLayer/pom.xml:20]: why not 17? :D

File: JpaLayer/pom.xml [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <version>4.13.2</version>
+      <scope>test</scope>
+    </dependency>
+
+    <dependency>
+      <groupId>org.springframework.boot</groupId>
+      <artifactId>spring-boot</artifactId>
```
Comment (nolik) [inline JpaLayer/pom.xml]: Not related to Kafka topic, but strongly recommend you to read about:
https://docs.spring.io/spring-boot/docs/1.2.3.RELEASE/reference/html/using-boot-build-systems.html#using-boot-maven-parent-pom
spring-boot-starter-parent and BOM (that able to take care for libs versioning).

File: JpaLayer/src/main/java/com/axinalis/config/KafkaConsumersConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    @Bean
+    ConcurrentKafkaListenerContainerFactory<String, String> kafkaListenerContainerFactory(){
+        ConcurrentKafkaListenerContainerFactory<String, String> factory =
+                new ConcurrentKafkaListenerContainerFactory<>();
+        factory.setConsumerFactory(consumerFactory());
+        return factory;
+    }
+
```
Comment (nolik) [inline JpaLayer/src/main/java/com/axinalis/config/KafkaConsumersConfiguration.java:51]: you even may specify custom Kafka templates with predefined Serializing / Deserializing:
https://www.baeldung.com/spring-kafka

File: JpaLayer/src/main/java/com/axinalis/consumer/TransactionsConsumer.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+                                @Autowired ObjectMapper mapper) {
+        this.repository = repository;
+        this.mapper = mapper;
+    }
+
+    @KafkaListener(topics = "transactions")
+    public void listen(Transaction message) {
+        TransactionEntity transactionToDb = new TransactionEntity();
+
+        transactionToDb.setBank(message.getBank());
```
Comment (nolik) [inline JpaLayer/src/main/java/com/axinalis/consumer/TransactionsConsumer.java]: Also could be off-top related to Kafka course: but I would recommend you to read about Layered Architecture and DAO layer (probably best option to move transaction persistence into DAO):
https://www.javaguides.net/2020/06/best-practice-to-develop-persistence-or-dao-layer.html
https://stackoverflow.com/questions/13785634/responsibilities-and-use-of-service-and-dao-layers

(for this particular course it's not so critical, but still recommend to check).

File: UserInterfaceLayer/src/main/java/com/axinalis/controller/UserInterfaceController.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+@RestController
+@RequestMapping("kafka")
+public class UserInterfaceController {
+
+    private KafkaTemplate<String, String> template;
+    private ClientsConsumer clientsConsumer;
+    private TransactionsConsumer transactionsConsumer;
+    @Value("${CREATE_USER_TOPIC}")
+    private String createUserTopic;
```
Comment (nolik) [inline UserInterfaceLayer/src/main/java/com/axinalis/controller/UserInterfaceController.java:22]: would recommend to move this topics into separate @Configuration.

File: JpaLayer/src/test/java/com/axinalis/JpaAppTest.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
+package com.axinalis;
+
+import junit.framework.Test;
+import junit.framework.TestCase;
+import junit.framework.TestSuite;
+
+
+public class JpaAppTest extends TestCase {
```
Comment (nolik) [inline JpaLayer/src/test/java/com/axinalis/JpaAppTest.java]: pretty dummy test. should it be some context test?

File: UserInterfaceLayer/src/main/java/com/axinalis/controller/UserInterfaceController.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    @Value("${TRANSACTIONS_TOPIC}")
+    private String transactionsTopic;
+    private ObjectMapper mapper;
+
+    public UserInterfaceController(@Autowired KafkaTemplate<String, String> template,
+                                   @Autowired ClientsConsumer clientsConsumer,
```
Comment (nolik) [inline UserInterfaceLayer/src/main/java/com/axinalis/controller/UserInterfaceController.java:32]: to simplify, may propose you to create dedicated KafkaService that will handle all Kafka communications.

File: JpaLayer/pom.xml (hunk 5 above)
Comment (Axinalis) [reply JpaLayer/pom.xml:20]: Java 11 has lighter images in docker
It's just for my comfort in this project

File: JpaLayer/src/test/java/com/axinalis/JpaAppTest.java (hunk 10 above)
Comment (Axinalis) [reply JpaLayer/src/test/java/com/axinalis/JpaAppTest.java]: It was created by maven
I'm just not sure how and what I should test in this module

File: JpaLayer/pom.xml (hunk 5 above)
Comment (nolik) [reply JpaLayer/pom.xml:20]: IMO java17 new LTS (be ready use new java features). in Docker you use it as runtime btw.

File: UserInterfaceLayer/UserInterfaceLayer.iml [hunk 12]
```diff
@@ -1,5 +1,18 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <module org.jetbrains.idea.maven.project.MavenProjectsManager.isMavenModule="true" type="JAVA_MODULE" version="4">
+  <component name="FacetManager">
```
Comment (nolik) [inline UserInterfaceLayer/UserInterfaceLayer.iml]: please, add .iml files to gitignore as well.

**Pull request merged**